libfprint*: update to 1.94.7

deviant opened the pull request:
## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Some of the libfprint-derived packages were getting quite out of date. I've updated everything but the non-free TOD drivers, and cleaned things up quite a bit.

Supersedes #290242 and #144843. Presumably fixes #133912 if it's still causing issues.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

deviant committed:
libfprint-tod: 1.90.7+git20210222+tod1 -> 1.94.7+tod1

Change-Id: Ib165ac807b3d977518d76e64ab7e78fc9e104dac

deviant committed:
fprintd-tod: rewrite package to follow upstream

Sharing logic from the base package and overriding where appropriate is
a better solution. As a side-effect, this brings the package up-to-date,
as it had been stuck at 1.90.9 for quite a while.

Change-Id: I2f33d690f4849a3225a16ca57d2945ef719db421

deviant committed:
libfprint-2-tod1-vfs0090: fix compilation with libfprint-tod 1.92.0+

Some APIs were changed in this version, in an incompatible fashion.

Change-Id: I9f312a54329c1f1877ddf451587ddf4fca9aabaf

deviant committed:
libfprint-2-tod1-vfs0090: remove libfprint dependency

This is not actually used by the package. Only the libfprint-tod fork is.

Change-Id: Id072b4673d2864e763f5149bdc2c58015c0d2fbb

deviant committed:
libfprint: 1.94.6 -> 1.94.7

Change-Id: Ibfafd64c95ac1a0c9355634dad2737d19c5fd21c

hmenke commented on pkgs/development/libraries/libfprint-tod/default.nix:
```diff
@@ -5,33 +5,22 @@
+libfprint.overrideAttrs (old: rec {
   pname = "libfprint-tod";
-  inherit version;
+  version = "1.94.7+tod1";
```
Newer versions of libfprint+tod have problems with enrolling fingerprints and upstream refuses to fix this issue, so we can never update beyond `1.90.7+git20210222+tod1`.

https://github.com/NixOS/nixpkgs/issues/175212#issuecomment-1175119104
https://gitlab.freedesktop.org/libfprint/libfprint/-/issues/484

deviant replied:
That's unfortunate! I think what I'll do is create a second, version-pinned `libfprint-tod`, so we can continue updating it (`1.90.7+git20210222+tod1` is 3 years old at this point, and has at least one serious issue for me that's been fixed since— it breaks after a single fingerprint attempt rather than allowing me to try authenticating again), whilst still allowing users of buggy proprietary drivers (which the `libfrpint-tod` author understandably cannot fix) to use their hardware.

deviant replied:
Done this now. ^^

deviant committed:
nixos/fprintd: use libfprint-tod_1_90 for proprietary drivers

These are either unknown to work or have been confirmed to not work
with the latest version of libfprint+tod. Given that they are not
particularly likely to be updated, and cannot be fixed due to their
closed nature, it's safest to pin them to v1.90 until proven to work.

Change-Id: I1c03985f67d0b5ceb005a5e6b4748ffc08199372

deviant committed:
libfprint-tod_1_90: init at 1.90.7+git20210222+tod1

Change-Id: Ia1c80ed2c8b51175e40df949bf61f7a9fa1aff30

hmenke commented on pkgs/development/libraries/libfprint-2-tod1-goodix/default.nix:
```diff
@@ -25,6 +25,7 @@ stdenv.mkDerivation {
   '';
 
   passthru.driverPath = "/lib/libfprint-2/tod-1";
+  passthru.needsLibfprint_1_90 = true;  # https://gitlab.freedesktop.org/libfprint/libfprint/-/issues/484
```
Are you certain that this is exhaustive? There are a few other Touch OEM Drivers. If this is not properly covering all the necessary ones, the next update of NixOS might break people's ability to enroll fingerprints or maybe even to login using fingerprint.

deviant replied:
I'm not at all sure this is exhaustive: I only added this for the one driver you gave me an issue link for. I don't have hardware for the other drivers, and so cannot test them (the linked issue details a runtime error). Do you think that pinning the version for all drivers unless proven to work would be a better idea? Perhaps just for the proprietary ones? Interested in your thoughts.

hmenke replied:
I think the other way around would be safer, although I admit this is quite conservative. At the same time, I would not want to lock someone out of their system, because they rely on fingerprint auth and forgot their password or something like that, even though that's obviously on them.

deviant replied:
Alright, let's do that! :) I'll add the attribute for all drivers other than `vfs0090`. I've not tested that one, but I am using [a fork of it](https://gitlab.com/bingch/libfprint-tod-vfs0090) and have pulled in a patch from that which makes it compile on v1.92.0+, so it seems relatively likely that it'll work.

Besides that, do you have any idea what the upgrade path for this will look like? How do we best get users to test if the latest version works? I'd like to pave the way to us either having a bug report link attached to each of the pins, or the pin removed if determined to not be necessary.

deviant replied:
Pushed. Given that `vfs0900` is both the only driver that didn't have this pinned, and is the only non-proprietary driver that we package, I'm less concerned about collecting bug reports for the rest. I'm not going to hold my breath for them to receive updates. And I suppose maintainers (or adventurous users) of the individual drivers may come along and check— I think that the attribute being visible in the module's `defaultText` will go some way towards piquing their interest/raising awareness.

hmenke replied:
How about guarding it by a [`stateVersion`](https://nixos.org/manual/nixos/stable/options#opt-system.stateVersion) check? I believe that is how the postgresql module handles possibly incompatible version upgrades.

deviant replied:
That would have to go purely in the module, so either a second attribute would have to be added, or the attribute would have to accept more states than just present/missing (e.g. present, missing, and "unknown"). That, plus the logic in the NixOS module would have to be made more complicated, to handle this additional condition. This isn't a stopper of course, but it's something to think about.

deviant commented:
Fixed a merge conflict.

SuperSandro2000 commented on pkgs/tools/security/fprintd/tod.nix:
```diff
@@ -1,51 +1,37 @@
-      maintainers = with lib.maintainers; [ hmenke ];
-    };
-  })
+  meta = {
```
Is this recursively updated, as in we retain the other meta attributes?

SuperSandro2000 commented on pkgs/development/libraries/libfprint-2-tod1-goodix/default.nix:
```diff
@@ -25,6 +25,7 @@ stdenv.mkDerivation {
   '';
 
   passthru.driverPath = "/lib/libfprint-2/tod-1";
+  passthru.needsLibfprint_1_90 = true;  # https://gitlab.freedesktop.org/libfprint/libfprint/-/issues/484
```
```suggestion
  passthru = {
    driverPath = "/lib/libfprint-2/tod-1";
    # https://gitlab.freedesktop.org/libfprint/libfprint/-/issues/484
    needsLibfprint_1_90 = true;
  };
```
same for the other 2

nyabinary commented:
bump :p

deviant commented:
@nyabinary Hi! I'm currently in the process of switching away from NixOS, so I won't be working on this any more. Feel free to pick up where I've left off.
